doc/user: improve Kafka source and sink docs

benesch opened the pull request:
This commit tackles several long-desired improvements to the Kafka source and sink documentation:

  * A fuller description of how `FORMAT JSON` and `FORMAT AVRO` actually encode data for sinks.

  * A description of `ENVELOPE DEBEZIUM`, and guidance on when to choose `ENVELOPE UPSERT` vs `ENVELOPE DEBEZIUM`.

  * Guidance on how to deal with common upsert `KEY` errors, including use of the `NOT ENFORCED` escape hatch.

  * An enumeration of what permissions are required on the Kafka broker for both sources and sinks.

Fix #11487.
Fix #13119.
Fix #18848.
Fix #19143.
Fix #20183.

<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

  * This PR fixes several recognized bugs.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

def- commented on src/sql/src/plan/error.rs:
```diff
@@ -213,6 +218,31 @@ impl PlanError {
+                let valid_keys = if valid_keys.is_empty() {
+                    "There are no known valid unique keys for the underlying relation.".into()
+                } else {
+                    format!(
```
I think this would be interesting to cover, probably in test/testdrive/kafka-avro-upsert-sinks.td. I'll try to do that.

def- reviewed (approved):
After triggering [nightly](https://buildkite.com/materialize/nightlies/builds/2669) I noticed the failures were also happening in normal tests. I hope it's ok that I fixed them inline in this PR.

Coverage marked inline in the files: https://buildkite.com/materialize/coverage/builds/163

bkirwi commented on doc/user/content/sql/create-source/kafka.md:
```diff
@@ -298,6 +298,15 @@ As long as any offset continues increasing, Materialize is consuming data from
+
+Operations  | Resource type    | Resource name
+------------|------------------|--------------
+Read        | Topic            | The specified `TOPIC` option
```
The Kafka source will also attempt to commit offsets under the consumer group it creates, though IIRC it will fail open. We might want to include that permission here to avoid surprises later.

benesch replied:
Oh, great point, thanks!

bkirwi reviewed (approved):
One suggestion! Overall this seems accurate and very helpful!

jseldess commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -192,13 +414,38 @@ CREATE CONNECTION csr_basic_http
+
+#### Debezium envelope
 
 {{< tabs tabID="1" >}}
```
I don't *think* you need this?

benesch replied:
Good point. Removed.

jseldess commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -192,13 +414,38 @@ CREATE CONNECTION csr_basic_http
+  KEY (key_col)
+  FORMAT JSON
+  ENVELOPE DEBEZIUM
+  WITH (SIZE = '3xsmall');
```
Same comment as before. If we want to start discouraging linked clusters, we might want to use `IN CLUSTER` here.

benesch replied:
Done.

jseldess reviewed (approved):
These doc changes look amazing, @benesch! Left a few nits and comments.

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+<b>Syntax:</b> `FORMAT JSON`
+
+The JSON format encodes each Kafka message value as a JSON object containing a
+field for each column of the sinked relation. The names and ordering of the
```
Oh good catch, thanks!

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+The Avro format encodes each Kafka message value as an Avro record containing a
+field for each column of the sink's source relation. The names and ordering of
+the fields in the record match the names and ordering of the columns in the
+relation. SQL types are converted to Avro types according to the type conversion
```
Done.

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+
+<br>
+<details>
+<summary>Type conversion table</summary>
```
Done.

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+relation. SQL types are converted to Avro types according to the type conversion
+table below.
+
+If the `KEY` option is specified, the Avro format encodes each Kafka message key
```
Ah, buried in the paragraph is "Kafka message value" vs "Kafka message key". The `KEY` options additional key metadata to each message. I've reworded to try to make that more clear. What do you think of the new phrasing? I still don't love it, but I think it's better.

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+fields in the record match the names and ordering of the columns in the
+relation.
+
+If the `KEY` option is specified, the JSON format encodes each Kafka message key
```
(Same response as for Avro.)

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+  FROM deduped
+  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
+  KEY (k)
+  FORMAT JSON ENVELOPE UPSERT WITH (SIZE = '2xmsall');
```
Done.

benesch commented on doc/user/content/sql/create-sink/kafka.md:
````diff
@@ -69,33 +69,236 @@ Field                | Value  | Description
+The Debezium envelope wraps each event in an object containing a `before` and
+`after` field:
+
+```json
````
Done.

benesch replied:
Yeah, I wanted the JSON highlighting! GitHub can't handle the `//` comments and the `...` but our docs handle it nicely.

benesch commented:
Thanks for the great feedback, @jseldess! Pushed up a change that addresses most of your comments.

> One additional thought, @benesch. The format and envelope details for `CREATE SOURCE` are on the main page, not the Kafka-specific page, although those details are not at all relevant for Postgres. I would consider moving those details to the Kafka-specific `CREATE SOURCE` page, to mirror the approach you've taken for `CREATE SINK`.
> 
> Related issue: #19416.

Heh, back in the day Marta did exactly the opposite refactor (#7885). Having the source formats separate from the source connections made more sense when we had other source types (files, S3) that also supported JSON and Avro. I agree in the current world of supporting only Kafka and PostgreSQL that we'd be better served by having the format and envelope docs on the one source where they matter.

benesch commented:
> I hope it's ok that I fixed them inline in this PR.

Absolutely it is! Thanks so much, @def-!

benesch commented:
Oh, crap, there _were_ some docs for sink formats and envelopes on the top level `CREATE SINK` page that I was unaware of. I guess ... I'll just remove those? 🙈

sjwiesman commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,252 @@ Field                | Value  | Description
+
+### Avro
+
+<b>Syntax:</b> `FORMAT JSON`
```
```suggestion
<b>Syntax:</b> `FORMAT AVRO`
```

benesch replied:
Ugh, thanks, fixed!

sploiselle commented on doc/user/content/sql/create-source/kafka.md:
```diff
@@ -298,6 +298,67 @@ As long as any offset continues increasing, Materialize is consuming data from
 
+### Committing offsets
+
+Kafka sources commit offsets when the messages up through that offset have been
+durably recorded in Materialize's storage layer. This enables Kafka tools that
+use committed offsets to determine consumer lag to monitor Materialize's
+progress through the source's topic.
+
+Prefer using Materialize's native [progress
+monitoring](#monitoring-source-progress) to the committed offsets in Kafka. The
+progress information in Materialize will be more up to date than the committed
+offsets in Kafka.
```
```suggestion
To support Kafka tools that monitor consumer lag, Kafka sources commit offsets
once durably recording them in Materialize's storage layer.

However, rather than relying on committed offsets, Materialize suggests using
our native [progress monitoring](#monitoring-source-progress) which always
contains more up-to-date data.
```

benesch replied:
Thanks, accepted with a tiny modification.

sploiselle commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,252 @@ Field                | Value  | Description
-[//]: # "TODO(morsapaes) Add information about upsert key selection"
+Consider using the Debezium envelope if:
+
+  * You have downstream consumers that want update events to contain both the
```
nit weird to have a single-element bulleted list

benesch replied:
Oh good point. I missed the second bullet that I meant to add here ("there is no natural key for the sink").

sploiselle reviewed (commented):
Just a drive by

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+SQL values are converted to JSON values according to the following rules:
+
+  * [`boolean`] values are converted to `true` or `false`.
+  * [`smallint`], [`integer`], [`bigint`], [`uint2`], [`uint4`], [`uint8`],
+    [`real`], and [`double precision`] values are converted to JSON numbers.
+  * [`numeric`] values are converted to a JSON string containing the
+    decimal representation of the number.
+  * [`record`] types are converted to JSON objects. The names and ordering of
+    the fields in the object match the names and ordering of the fields in the
+    record.
+  * [`list`] values are converted to JSON arrays.
+  * [array][arrays] values are converted to JSON arrays. Multidimensional arrays
+    are flattened into a single dimension following row-major order.
+  * [`timestamp`] and [`timestamp with time zone`] values are converted to JSON
+    strings containing the number of milliseconds since the Unix epoch.
+  * All other types are cast to [`text`] and then converted to JSON strings.
```
```suggestion
SQL type                     | Conversion
-----------------------------|-------------------------------------
[`array`][`arrays`]          | Values are converted to JSON arrays. Multidimensional arrays are flattened into a single dimension following row-major order.
[`bigint`]                   | Values are converted to JSON numbers.
[`boolean`]                  | Values are converted to `true` or `false`.
[`integer`]                  | Values are converted to JSON numbers.
[`list`]                     | Values are converted to JSON arrays.
[`numeric`]                  | Values are converted to a JSON string containing the decimal representation of the number.
[`record`]                   | Records are converted to JSON objects. The names and ordering of the fields in the object match the names and ordering of the fields in the record.
[`smallint`]                 | values are converted to JSON numbers.
[`timestamp`][`timestamp with time zone`] | Values are converted to JSON strings containing the number of milliseconds since the Unix epoch.
[`uint2`]                    | Values are converted to JSON numbers.
[`uint4`]                    | Values are converted to JSON numbers.
[`uint8`]                    | Values are converted to JSON numbers.
Other                        | Values are cast to [`text`] and then converted to JSON strings.
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+## Envelopes
+
+The sink's envelope determines how inserts, updates, and delete events are
```
```suggestion
In addition to determining how to encode output data, Materialize also needs to
understand the desired behaviour of sinked events in the downstream external
system. Whether a change is simply emitted as an event with a diff structure,
or actively inserts, updates, or deletes existing data downstream depends on
the `ENVELOPE` specified in the `CREATE SINK` statement.
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+### Upsert
+
+<b>Syntax:</b> `ENVELOPE UPSERT`
```
```suggestion
<p style="font-size:14px"><b>Syntax:</b> <code>ENVELOPE UPSERT</code></p>
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+The upsert envelope:
+
+  * Emits insertion and update events without additional decoration.
+  * Converts deletion events into messages with a `null` value (often called
+    "tombstones").
```
```suggestion
  * Converts deletion events into messages with a `NULL` value (i.e. _tombstones_).
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+  * You want to enable key-based compaction on the sink's Kafka topic while
+    retaining the most recent value for each key.
+
+#### Key selection
```
The content in this section is very troubleshooting-focused, so we're again mixing and matching reference documentation and guides, but we can figure out how to move content around later.

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+  {{< note >}}
+  Maintaining the `deduped` materialized view requires memory proportional to the
+  number of records in `original_input`. Be sure to assign `deduped`
+  to a cluster with resources sufficient to handle your data volume.
```
```suggestion
  to a cluster with adequate resources to handle your data volume.
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+### Debezium
+
+<b>Syntax:</b> `ENVELOPE DEBEZIUM`
```
```suggestion
<p style="font-size:14px"><b>Syntax:</b> <code>ENVELOPE DEBEZIUM</code></p>
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+<b>Syntax:</b> `ENVELOPE UPSERT`
+
+The upsert envelope:
```
```suggestion
The upsert envelope treats all records as having a **key** and a **value**, and
supports inserts, updates and deletes in the sink:
```

benesch replied:
Similar thought to the above. It's hard for me to square "supporting updates and deletes" with what the upsert envelope actually does.

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+
+<b>Syntax:</b> `ENVELOPE DEBEZIUM`
+
+The Debezium envelope wraps each event in an object containing a `before` and
+`after` field to indicate whether the event was an insertion, deletion, or
+update:
```
```suggestion
The Debezium envelope wraps each event in an object containing a `before` and
`after` field to indicate whether the event was an insertion, deletion, or
update to the underlying source, table or materialized view being written to
the sink:
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -113,6 +334,25 @@ Exactly-once semantics are an end-to-end property of a system, but Materialize o
 
+## Required permissions
+
+The access control lists (ACLs) on the Kafka cluster must permit Materialize
```
```suggestion
The access control lists (ACLs) on the Kafka cluster must allow Materialize
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -113,6 +334,25 @@ Exactly-once semantics are an end-to-end property of a system, but Materialize o
+When using [automatic topic creation](#automatic-topic-creation), Materialize
+additionally requires access to the following operations:
+
+Operations       | Resource type    | Resource name
```
```suggestion
Operation type       | Resource type    | Resource name
```

morsapaes commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -113,6 +334,25 @@ Exactly-once semantics are an end-to-end property of a system, but Materialize o
+The access control lists (ACLs) on the Kafka cluster must permit Materialize
+to perform the following operations on the following resources:
+
+Operations  | Resource type    | Resource name
```
```suggestion
Operation type | Resource type    | Resource name
```

benesch commented:
Thanks for the review, @morsapaes. I've addressed almost all of your feedback, with just a few exceptions, mostly around envelopes. 

I'm going to take another pass at incorporating your suggestions into the envelope text, but I'd like to find a way to not hold up the PR over that because there's a number of other changes in here that are ready to go. Unfortunately it's hard to back just the envelope changes because they're tangled up with the key selection guidance etc. Do you have a preference between these options?

  1. Accept your suggestions verbatim, send up another PR with improvements.
  2. Merge as is, send up another PR with wording improvements.
  3. No it's really bad, keep workshopping in this PR.

morsapaes reviewed (approved):
No need to block the PR on the drive-by feedback. At this point, we need another makeover of the `CREATE SOURCE` and `CREATE SINK` pages, so we can fine tune things then.
